design/how-ironic-works.md

```diff
@@ -0,0 +1,356 @@
+
+In order to assert this configuration at boot time, a dedicated DHCP server
+on semi-dedicated "provisioning" network should be leveraged. This network
+may be reused by other services and systems, but can also be leveraged in
+the case of hardware discovery. The most important factor is that this
+network does not have an external DHCP server attached.
+
+Some specific hardware types in ironic do support use of virtual media
+to boot the deployment ramdisk, however this functionality is not
+available with vendor neutral IPMI. Newer protocols such as Redfish
+may support virtual media, but not as of the time of this document
+having been composed. Use of virtual media should be considered
+useful, but not well enough supported at this time to be considered useful.
+Other methods such as booting directly from iSCSI should be considered
+out-of-scope in this use case as they are centered around an external block
+storage management system.
+
+## How ironic writes an operating system image to baremetal?
+
+Ironic supports two fundamental types of disk images.
+
+Whole-disk and Partition (or filesystem) images.
+
+Due to the nature of partition images, where a bootloader and configuration
+must be put in place and configured, it is simpler to focus on the use of
+whole disk images.
+
+The basic workflow consists of:
+
+1. Booting the deployment ramdisk
+2. The deployment ramdisk checks in with the ironic introspection
+   service which updates configuration information stored in ironic
+   about the baremetal machine.
+3. The deployment ramdisk checks in with the ironic service and leverages
+   the MAC addresses to help identify the physical machine in the hardware
+   inventory.
+4. Ironic initiates deployment by first identifying the root disk upon
+   which the disk image is to be written. By default this will be the
+   smallest storage device available, and can be overridden via explict
+   configuration of a [root_device hint][https://docs.openstack.org/ironic/latest/install/advanced.html#specifying-the-disk-for-deployment-root-device-hints]
+5. The deployment ramdisk downloads the image to be written and streams
+   that to the storage device.
+6. If defined as part of the deployment, ironic will add an additional
+   partition for a configuration drive. Ironic will then write the
+   configuraiton drive to disk
+7. Finally ironic reboots the machine.
+
+There are some additional steps that ironic performs, mainly fixing
+partition table data with GPT based partition tables in order to
+prevent issues after deployment, but these steps are incorporated
+as part of the deployment sequence to help ensure that the machine
+will deploy successfully without issues.
+
+## What connectivity is required?
+
+To boot the discovery and deployment image on the baremetal hardware:
+
+* DHCP
+* TFTP (if iPXE is not natively supported by the network interfaces.)
+* HTTP in order to download kernel/ramdisk images via HTTP over a TCP
+  connection.
```

Comment:
All 3 of those connections would be back to the place where Ironic is running, right?

Reply:
Yes. Do we want to revise this to stress this further?

Reply:
I'll propose an update after this merges.